crates/task-impls/src/quorum_proposal/handlers.rs

```diff
@@ -0,0 +1,332 @@
+
+                    leaf.fill_block_payload_unchecked(payload);
+                }
+
+                let vid_share = vid_shares
+                    .get(&leaf.get_view_number())
+                    .unwrap_or(&HashMap::new())
+                    .get(&task_state.public_key)
+                    .cloned()
+                    .map(|prop| prop.data);
+
+                // Add our data into a new `LeafInfo`
+                ret.leaf_views.push(LeafInfo::new(
+                    leaf.clone(),
+                    Arc::clone(&state),
+                    delta.clone(),
+                    vid_share,
+                ));
+                ret.leaves_decided.push(leaf.clone());
+                if let Some(ref payload) = leaf.get_block_payload() {
+                    for txn in payload.transaction_commitments(leaf.get_block_header().metadata()) {
+                        ret.included_txns.insert(txn);
+                    }
+                }
+            }
+        } else {
+            bail!(
+                    "Validated state and delta do not exist for the leaf for view {current_leaf_view_number:?}"
+                )
+        };
+
+        // Move on to the next leaf at the end.
+        next_leaf = leaf.get_parent_commitment();
+    }
+
+    bail!("Leaf not found");
+}
+
+/// Handles the `QuorumProposalValidated` event.
+pub(crate) async fn handle_quorum_proposal_validated<
+    TYPES: NodeType,
+    I: NodeImplementation<TYPES>,
+>(
+    proposal: &QuorumProposal<TYPES>,
+    sender: &Sender<Arc<HotShotEvent<TYPES>>>,
+    task_state: &mut QuorumProposalTaskState<TYPES, I>,
+) -> Result<()> {
+    let LeafChainTraversalOutcome {
+        new_locked_view_number,
+        new_decided_view_number,
+        new_decide_qc,
+        leaf_views,
+        leaves_decided,
+        included_txns,
+    } = visit_leaf_chain(proposal, task_state).await?;
+
+    let included_txns = if new_decided_view_number.is_some() {
+        included_txns
+    } else {
+        HashSet::new()
+    };
+
+    let mut consensus_writer = task_state.consensus.write().await;
+    if let Some(locked_view_number) = new_locked_view_number {
+        // Broadcast the locked view update.
+        broadcast_event(
+            HotShotEvent::LockedViewUpdated(locked_view_number).into(),
+            sender,
+        )
+        .await;
+
+        consensus_writer.update_locked_view(locked_view_number)?;
+    }
+
+    // TODO - update decided upgrade cert
+
+    #[allow(clippy::cast_precision_loss)]
+    if let Some(decided_view_number) = new_decided_view_number {
+        // Bring in the cleanup crew. When a new decide is indeed valid, we need to clear out old memory.
+
+        let old_decided_view = consensus_writer.last_decided_view();
+        // TODO - collect garbage.
+
+        // Set the new decided view.
+        consensus_writer.update_last_decided_view(decided_view_number)?;
```

Comment:
Shouldn't we also send `LastDecidedViewUpdated`?

Reply:
Correct, fixed.